### PR TITLE
Upgrade to Omnisharp v1.38.0

### PR DESCRIPTION
**Pull request by rita-sonarsource**

### CODE REVIEW

File: omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj [hunk 1]
```diff
@@ -1,30 +1,30 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
-  <PropertyGroup>
-    <TargetFramework>.netcoreapp3.1</TargetFramework>
-    <LangVersion>9</LangVersion>
-    <!-- We need the OmniSharp assemblies to execute the tests-->
-    <CopyOmniSharpToOutput>true</CopyOmniSharpToOutput>
-  </PropertyGroup>
+	<PropertyGroup>
+		<TargetFramework>net472</TargetFramework>
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj:4]: The latest version of OmniSharp references `Microsoft.CodeAnalysis.Workspaces` 4.0.0.0, which conditionally includes System.Range _if_ the target framework is _not_ .net core. 

image

Hence, if we're building on .net core, the class System.Range would not be there, as it's supposed to be part of the .net core framework.

It should've just worked but in practice I was getting the exception
```
System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
Could not load type 'System.Range' from assembly 'Microsoft.CodeAnalysis.Workspaces, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
    at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.RuntimeAssembly.get_DefinedTypes()
   at System.Composition.Hosting.ContainerConfiguration.<>c.<WithAssemblies>b__16_0(Assembly a)
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at System.Composition.TypedParts.TypedPartExportDescriptorProvider..ctor(IEnumerable`1 types, AttributedModelProvider attributeContext)
   at System.Composition.Hosting.ContainerConfiguration.CreateContainer()
   at Microsoft.CodeAnalysis.Host.Mef.MefHostServices.Create(IEnumerable`1 assemblies)
   at OmniSharp.HostServicesAggregator.CreateHostServices() in /home/vsts/work/1/s/src/OmniSharp.Roslyn/HostServicesAggregator.cs:line 61
   at OmniSharp.OmniSharpWorkspace..ctor(HostServicesAggregator aggregator, ILoggerFactory loggerFactory, IFileSystemWatcher fileSystemWatcher) in /home/vsts/work/1/s/src/OmniSharp.Roslyn/OmniSharpWorkspace.cs:line 53
   at SonarLint.OmniSharp.DotNet.Services.UnitTests.DiagnosticWorker.SonarLintDiagnosticWorkerTests.CreateOmniSharpWorkspace() in C:\Projects\sonarlint-omnisharp\omnisharp-dotnet\src\Services.UnitTests\DiagnosticWorker\SonarLintDiagnosticWorkerTests.cs:line 213
   at SonarLint.OmniSharp.DotNet.Services.UnitTests.DiagnosticWorker.SonarLintDiagnosticWorkerTests.CreateOmnisharpWorkspaceWithDocument(String documentFileName, String documentContent) in C:\Projects\sonarlint-omnisharp\omnisharp-dotnet\src\Services.UnitTests\DiagnosticWorker\SonarLintDiagnosticWorkerTests.cs:line 205
   at SonarLint.OmniSharp.DotNet.Services.UnitTests.DiagnosticWorker.SonarLintDiagnosticWorkerTests.AnalyzeDocument_AnalyzerOptionsAreOverriden() in C:\Projects\sonarlint-omnisharp\omnisharp-dotnet\src\Services.UnitTests\DiagnosticWorker\SonarLintDiagnosticWorkerTests.cs:line 133
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadOperations.ExecuteWithAbortSafety(Action action)
```

Changing it to net472 _or_ .net6 solved the problem. I had other problems changing it to .net6 on the omnisharp side, so for now I went with 472. I think it's ok as it matches the version of OmniSharp that we import at the moment.

File: omnisharp-dotnet/src/Services/SonarLint.OmniSharp.DotNet.Services.csproj [hunk 2]
```diff
@@ -1,59 +1,60 @@
...
-    <PackageReference Include="Microsoft.Extensions.Options" Version="3.1.12" />
-  </ItemGroup>
-  
-  
-  <!-- ********************************************** -->
-  <!-- Packaging files to be included in the Java jar -->
-  <!-- ********************************************** -->
-    <!-- Add any files that should be included in the Java plugin to the item group @"FilesToIncludeInJar".
+    -->
+		<PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.0.0-6.final" />
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services/SonarLint.OmniSharp.DotNet.Services.csproj:15]: Versions that match OmniSharp + additional DLLs that the code was failing to find.

File: omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj [hunk 3]
```diff
@@ -1,30 +1,39 @@
...
+	<ItemGroup>
+		<PackageReference Include="coverlet.collector" Version="3.0.3">
+			<PrivateAssets>all</PrivateAssets>
+			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+		</PackageReference>
+		<PackageReference Include="coverlet.msbuild" Version="3.0.3">
+			<PrivateAssets>all</PrivateAssets>
+			<IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+		</PackageReference>
+		<PackageReference Include="FluentAssertions" Version="6.3.0" />
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj:19]: I had a lot of problems comparing the XML strings in SonarLintConfigurationSerializerTests. Upgrading to the latest version solved it.

File: omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj [hunk 4]
```diff
@@ -1,30 +1,39 @@
...
+		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.6.1" />
+		<PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0">
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj:21]: Was needed on the CI, https://andrewlock.net/using-reference-assemblies-to-build-net-framework-libararies-on-linux-without-mono/

File: omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj [hunk 5]
```diff
@@ -1,30 +1,39 @@
...
+		  <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+		</PackageReference>
+		<PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies.net472" Version="1.0.0">
+		  <PrivateAssets>all</PrivateAssets>
+		  <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+		</PackageReference>
+		<PackageReference Include="Moq" Version="4.14.5" />
+		<PackageReference Include="MSTest.TestAdapter" Version="1.4.0" />
+		<PackageReference Include="MSTest.TestFramework" Version="1.4.0" />
+		<PackageReference Include="SQLitePCLRaw.core" Version="2.0.4" />
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services.UnitTests/SonarLint.OmniSharp.DotNet.Services.UnitTests.csproj:32]: Used by OmniSharp

File: omnisharp-dotnet/src/Services.UnitTests/Rules/SonarLintXml/RulesToSonarLintConfigurationConverterTests.cs [hunk 6]
```diff
@@ -75,12 +75,18 @@ public void Convert_RulesWithParametersAreIncluded()
 
             sonarLintConfiguration.Rules[0].Key.Should().Be("rule1");
             sonarLintConfiguration.Rules[0].Parameters.Should().BeEquivalentTo(
-                new SonarLintKeyValuePair{Key = "param1", Value = "value1"});
+                new[]
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services.UnitTests/Rules/SonarLintXml/RulesToSonarLintConfigurationConverterTests.cs:78]: Minor: changes in the latest fluent assertions, now it requires to pass a collection when comparing

File: omnisharp-dotnet/src/Services.UnitTests/DiagnosticWorker/SonarLintDiagnosticWorkerTests.cs [hunk 7]
```diff
@@ -142,7 +142,7 @@ AnalyzerOptions ModifyOptions(AnalyzerOptions originalOptions)
             // second call should have an additional file
             result = await testSubject.AnalyzeDocumentAsync(document, CancellationToken.None);
             result.FirstOrDefault(x => x.Id == TestAnalyzer.Descriptor.Id).Should().NotBeNull();
-            analyzer.SuppliedAdditionalFiles.Should().BeEquivalentTo(additionalFile);
+            analyzer.SuppliedAdditionalFiles.Should().BeEquivalentTo(new [] {additionalFile});
```
Comment (rita-sonarsource) [inline omnisharp-dotnet/src/Services.UnitTests/DiagnosticWorker/SonarLintDiagnosticWorkerTests.cs:145]: Minor: changes in the latest fluent assertions, now it requires to pass a collection when comparing

**Review by duncanp-sonar — APPROVED** [review]

**Pull request merged**